### PR TITLE
Feature/#62 trade

### LINKED ISSUE #62: feature/#62 - Trade Implementation
Add base trade implementation for animal exchange.

### PR DESCRIPTION
**Pull request by Enessetere**
closes #62 

demo in manuals

### CODE REVIEW

File: src/app/BreedProcessor.ts [hunk 1]
```diff
@@ -6,91 +6,110 @@ import { AnimalNames } from '../Enums/AnimalNamesEnum';
...
+    herd.theAnimals
+      .filter(
+        (tuple) =>
+          tuple[0].hasRole(AnimalRoles.LIVESTOCK) &&
+          tuple[0].theName !== AnimalNames.HORSE,
+      )
+      .forEach((tuple) =>
+        this.bank.theHerd.addAnimals(
+          tuple[0].theName as AnimalNames,
+          tuple[1],
```
Comment (DominikNowak) [inline src/app/BreedProcessor.ts]: Consider use of destructuring to increase clarity

File: src/app/Trade.ts [hunk 2]
```diff
@@ -0,0 +1,70 @@
+import { AnimalNames } from '../Enums/AnimalNamesEnum';
+import { ConvertToAnimalObject } from './utils/ConvertToAnimalObject';
+import _ from 'lodash';
```
Comment (DominikNowak) [inline src/app/Trade.ts]: There is only multiply used here from lodash, so maybe just { multiply } ?

**Review by DominikNowak — DISMISSED** [review]
There are few places where you can use destructuring, but it generally looks fine. Well done :)

File: src/app/BreedProcessor.ts (hunk 1 above)
Comment (Enessetere) [reply src/app/BreedProcessor.ts]: good idea. added to refactor list.

File: src/app/Trade.ts (hunk 2 above)
Comment (Enessetere) [reply src/app/Trade.ts]: right. forgot to change it. added to refactor list :)

File: src/app/BreedProcessor.ts (hunk 1 above)
Comment (Enessetere) [reply src/app/BreedProcessor.ts]: fixed

File: src/app/Trade.ts (hunk 2 above)
Comment (Enessetere) [reply src/app/Trade.ts]: fixed

File: src/app/Trade.ts [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+
+  private adjustOffer(
+    offer: [AnimalNames, number],
+    target: [AnimalNames, number],
+  ): void {
+    offer[1] -= 1;
+    if (this.calculateValue(offer) <= this.calculateValue(target)) {
+      return;
+    }
+    this.adjustOffer(offer, target);
```
Comment (synowa) [inline src/app/Trade.ts:53]: IMO loop would be more readable here.

File: src/app/Trade.ts [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+  }
+
+  /**
+   * updates players and banks herd
+   */
+  private disposeResult(
```
Comment (synowa) [inline src/app/Trade.ts:59]: I'm not sure what disposeResult means, maybe executeTrade/ performTrade would be clearer?

File: src/app/Trade.ts [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+   * Gets an offer from player and returns true or false if transaction can be made processed and process it if possible
+   * @param offer accepts tuple with offer containing animal name and quantity to be sold
+   * @param player accepts instance of Player class, which wants to sell his animals
+   * @param target accepts tuple with desired animal(s) containing desired animal name and quantity to buy
+   * @returns true if transaction will be processed, and false otherwise
+   */
+  processOffer(
+    offer: [AnimalNames, number],
+    { theHerd: playerHerd }: Player,
+    target: [AnimalNames, number],
```
Comment (synowa) [inline src/app/Trade.ts:19]: Great idea with destructuring first and third argument in disposeResult method, maybe the same could be applied here.

**Review by synowa — APPROVED** [review]

**Review by SebastianBabinski1 — APPROVED** [review]
Wow, you did a great job! Looks fine, just Trade.ts was hard to understand for me :)

**Pull request merged**